### PR TITLE
feat: [IDP-1278]: Audit trial setup for IDP - AppConfigs and env variables

### PR DESCRIPTION
**Pull request by devesh-harness**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53509)

### CODE REVIEW

**Review by vikyathharekal — COMMENTED** [review]
> Reviewed 10 of 20 files at r1, 1 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53509)* status: 20 file(s) to review. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/audittrials/eventhandlers/AppConfigEventHandler.java` line 37 at r2](https://reviewable.io/reviews/harness/harness-core/53509#-NfRL6hWETTRieFEgn1l:-NfRL6hWETTRieFEgn1m:bfza8df) ([raw file](https://github.com/harness/harness-core/blob/e25f48161c24396c67ba860b2b2400000d7fd1dc/idp-service/src/main/java/io/harness/idp/audittrials/eventhandlers/AppConfigEventHandler.java#L37)):*
> ```Java
> @Slf4j
> public class AppConfigEventHandler implements OutboxEventHandler {
>   private final ObjectMapper objectMapper;
> ```

Can we make this static and init with NG_DEFAULT_OBJECT_MAPPER here itself?



_Code quote:_
```Java
objectMapper
```
___
*[`idp-service/src/main/java/io/harness/idp/audittrials/eventhandlers/AppConfigEventHandler.java` line 50 at r2](https://reviewable.io/reviews/harness/harness-core/53509#-NfRMjA1BEG8YYkbLapQ:-NfRMjA1BEG8YYkbLapR:b-uamsmn) ([raw file](https://github.com/harness/harness-core/blob/e25f48161c24396c67ba860b2b2400000d7fd1dc/idp-service/src/main/java/io/harness/idp/audittrials/eventhandlers/AppConfigEventHandler.java#L50)):*
> ```Java
>     try {
>       switch (outboxEvent.getEventType()) {
>         case APP_CONFIG_UPDATE:
> ```

nit: We could keep this consistent in terms of past/present tense. Since the actual DB operation has already happened xxx_CREATED/xxx_UPDATED might be better everywhere



_Code quote:_
```Java
APP_CONFIG_UPDATE
```
___
*[`idp-service/src/main/java/io/harness/idp/audittrials/eventhandlers/AppConfigEventHandler.java` line 52 at r2](https://reviewable.io/reviews/harness/harness-core/53509#-NfRM5gH4sMkbb8bq_99:-NfRM5gIEu19x3rUzrRD:b2oefay) ([raw file](https://github.com/harness/harness-core/blob/e25f48161c24396c67ba860b2b2400000d7fd1dc/idp-service/src/main/java/io/harness/idp/audittrials/eventhandlers/AppConfigEventHandler.java#L52)):*
> ```Java
>         case APP_CONFIG_UPDATE:
>           return handleAppConfigUpdateEvent(outboxEvent);
>         case APP_CONFIG_SAVED:
> ```

How about APP_CONFIG_CREATE? save could mean both create or update.



_Code quote:_
```Java
APP_CONFIG_SAVED
```
___
*[`idp-service/src/main/java/io/harness/idp/audittrials/eventhandlers/IDPNextGenOutboxEventHandler.java` line 36 at r2](https://reviewable.io/reviews/harness/harness-core/53509#-NfRNCEI7-d6sX1fyksC:-NfRNCEI7-d6sX1fyksD:bnjkmwg) ([raw file](https://github.com/harness/harness-core/blob/e25f48161c24396c67ba860b2b2400000d7fd1dc/idp-service/src/main/java/io/harness/idp/audittrials/eventhandlers/IDPNextGenOutboxEventHandler.java#L36)):*
> ```Java
>       log.error(
>           String.format("Unexpected error occurred during handling event of type %s", outboxEvent.getEventType()));
>       log.error("Exception: " + exception.toString());
> ```

nit: toString() is redundant

**Review by vikyathharekal — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53509)* status: 20 file(s) to review. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/configmanager/service/ConfigEnvVariablesServiceImpl.java` line 59 at r2](https://reviewable.io/reviews/harness/harness-core/53509#-NfRNlE84cjdweZD85bl:-NfRNlE84cjdweZD85bm:b-to4vin) ([raw file](https://github.com/harness/harness-core/blob/e25f48161c24396c67ba860b2b2400000d7fd1dc/idp-service/src/main/java/io/harness/idp/configmanager/service/ConfigEnvVariablesServiceImpl.java#L59)):*
> ```Java
>   @Inject @Named(OUTBOX_TRANSACTION_TEMPLATE) private TransactionTemplate transactionTemplate;
>   @Inject private OutboxService outboxService;
>   private final RetryPolicy<Object> transactionRetryPolicy = DEFAULT_RETRY_POLICY;
> ```

Can we make this private static final?



_Code quote:_
```Java
 private final
```

File: idp-service/src/main/java/io/harness/idp/audittrials/eventhandlers/dtos/BackstageEnvSecretDTO.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import lombok.Builder;
+import lombok.Data;
+import lombok.experimental.FieldDefaults;
+
+@Data
+@JsonIgnoreProperties(ignoreUnknown = true)
+@FieldDefaults(level = AccessLevel.PRIVATE)
+@Builder
+public class BackstageEnvSecretDTO implements YamlDTO {
+  String EnvVariableName;
```
[the remark this answers is not in the document]
Comment (devesh-harness) [reply idp-service/src/main/java/io/harness/idp/audittrials/eventhandlers/dtos/BackstageEnvSecretDTO.java:22]: done

**Comment by devesh-harness** [comment]
trigger uts

File: idp-service/src/main/java/io/harness/idp/app/IdpModule.java [hunk 2]
```diff
@@ -625,4 +644,18 @@ public IteratorConfig scorecardScoreComputationIteratorConfig() {
   public String idpServiceSecret() {
     return this.appConfig.getIdpServiceSecret();
   }
+
+  @Provides
```
Comment (sathish-soundarapandian) [inline idp-service/src/main/java/io/harness/idp/app/IdpModule.java:648]: nit: Do we need these two ? We should be able to access it directly itself

**Review by sathish-soundarapandian — APPROVED** [review]

**Review by vikyathharekal — APPROVED** [review]
> Reviewed 3 of 20 files at r1, 1 of 2 files at r2, 9 of 9 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53509)* status: 20 file(s) to review. 1. unresolved discussions. 

___
*[`idp-service/src/main/java/io/harness/idp/configmanager/events/AppConfigSaveEvent.java` line 29 at r3](https://reviewable.io/reviews/harness/harness-core/53509#-NfU5cg02ftsPkuCOG5Z:-NfU5cg02ftsPkuCOG5_:bpwp1iz) ([raw file](https://github.com/harness/harness-core/blob/21282755e95d776e3959555e139eee2a1e22fd7d/idp-service/src/main/java/io/harness/idp/configmanager/events/AppConfigSaveEvent.java#L29)):*
> ```Java
> @Getter
> @NoArgsConstructor
> public class AppConfigSaveEvent implements Event {
> ```

nit: AppConfigCreatedEvent



_Code quote:_
```Java
AppConfigSaveEvent
```

**Comment by devesh-harness** [comment]
trigger uts

**Comment by devesh-harness** [comment]
trigger uts

**Review by sathish-soundarapandian — APPROVED** [review]

**Review by vikyathharekal — APPROVED** [review]
> Reviewed 8 of 8 files at r4, 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53509)* status: 22 file(s) to review. 1. unresolved discussions.

**Comment by devesh-harness** [comment]
trigger ut4, ut7

**Comment by devesh-harness** [comment]
trigger ut4

**Comment by devesh-harness** [comment]
trigger ut4

**Comment by devesh-harness** [comment]
trigger uts

**Comment by devesh-harness** [comment]
trigger ut5

**Review by vikyathharekal — APPROVED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53509)* status: 22 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

File: idp-service/src/main/java/io/harness/idp/audittrials/eventhandlers/BackstageSecretEnvEventHandler.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+  @Inject
+  public BackstageSecretEnvEventHandler(AuditClientService auditClientService) {
+    this.auditClientService = auditClientService;
+  }
+
+  @Override
+  public boolean handle(OutboxEvent outboxEvent) {
+    try {
+      switch (outboxEvent.getEventType()) {
```
[the remark this answers is not in the document]
Comment (devesh-harness) [reply idp-service/src/main/java/io/harness/idp/audittrials/eventhandlers/BackstageSecretEnvEventHandler.java:46]: yes in the current flow that is not possible will add that as enhancement.

**Review by sathish-soundarapandian — APPROVED** [review]

**Comment by devesh-harness** [comment]
trigger review

**Pull request merged**